File: envdiff/render/human.py
from __future__ import annotations

from collections import Counter

from envdiff.models import Finding, RepoScanResult


def render_compare_result(result: dict[str, object]) -> str:
    lines = [
        f"Compare {result['left_path']} vs {result['right_path']}",
        f"Missing in left: {', '.join(result['missing_in_left']) or '-'}",
        f"Missing in right: {', '.join(result['missing_in_right']) or '-'}",
        f"Duplicates in left: {', '.join(result['duplicates_in_left']) or '-'}",
        f"Duplicates in right: {', '.join(result['duplicates_in_right']) or '-'}",
    ]

    differing = result["differing_values"]
    if differing:
        lines.append("Differing values:")
        for item in differing:
            lines.append(
                "  "
                f"{item['name']}: {item['left_kind']}={item['left_value']} vs "
                f"{item['right_kind']}={item['right_value']}"
            )

    return "\n".join(lines)


def render_scan_result(scan_result: RepoScanResult) -> str:
    lines = [
        f"Scan root: {scan_result.root_path}",
        f"Definitions: {len(scan_result.definitions)}",
        f"Usages: {len(scan_result.usages)}",
        f"Contracts: {len(scan_result.contracts)}",
    ]

    if scan_result.contracts:
        lines.append("Contracts:")
        for contract in scan_result.contracts:
            statuses = ",".join(contract.status) or "none"
            lines.append(f"  {contract.name} [{contract.requiredness}] ({statuses})")

    return "\n".join(lines)


def render_matrix_result(result: dict[str, object]) -> str:
    lines = [
        f"Matrix files: {result['file_count']}",
        f"Variables: {result['variable_count']}",
        f"Inconsistent: {result['inconsistent_variable_count']}",
    ]

    warnings = result["warnings"]
    if warnings:
        lines.append(f"Warnings: {len(warnings)}")

    variables = result["variables"]
    if not variables:
        lines.append("No variables matched the selected view.")
        return "\n".join(lines)

    lines.append("Variables:")
    for variable in variables:
        reasons = []
        if variable["missing_in"]:
            reasons.append("missing")
        if len(variable["value_kinds"]) > 1:
            reasons.append("kind-mismatch")
        if variable["duplicates_in"]:
            reasons.append("duplicate")
        lines.append(f"  {variable['name']} [{', '.join(reasons) or 'consistent'}]")
        for file_entry in variable["files"]:
            if file_entry["presence"] == "missing":
                lines.append(f"    {file_entry['path']}: missing")
                continue

            suffix = file_entry["value_kind"]
            if file_entry.get("is_duplicate"):
                suffix = f"{suffix}, duplicate"
            lines.append(f"    {file_entry['path']}: present ({suffix})")

    return "\n".join(lines)


def render_generate_result(
    variable_count: int,
    *,
    output_path: str | None = None,
    annotate: bool,
    check_path: str | None = None,
    check_matches: bool | None = None,
) -> str:
    if check_path is not None and check_matches is not None:
        if check_matches:
            return f"Generated output matches {check_path}"
        return f"Generated output drifted from {check_path}"

    suffix = " with annotations" if annotate else ""
    return (
        f"Generated {variable_count} variables{suffix} to {output_path}"
    )


def render_doctor_result(
    root_path: str,
    findings: tuple[Finding, ...],
    *,
    suppressed_count: int = 0,
    baseline_written: str | None = None,
) -> str:
    counts = Counter(finding.severity for finding in findings)
    lines = [
        f"Doctor root: {root_path}",
        f"Findings: {len(findings)}",
        (
            "Summary: "
            f"{counts.get('error', 0)} error, "
            f"{counts.get('warning', 0)} warning, "
            f"{counts.get('info', 0)} info"
        ),
    ]
    if suppressed_count:
        lines.append(f"Suppressed: {suppressed_count}")
    if baseline_written:
        lines.append(f"Baseline written: {baseline_written}")
    if not findings:
        lines.append("No active findings.")
        return "\n".join(lines)

    for severity in ("error", "warning", "info"):
        scoped = [finding for finding in findings if finding.severity == severity]
        if not scoped:
            continue
        lines.append(f"{severity.capitalize()}s:")
        for finding in scoped:
            lines.append(f"  {finding.code} {finding.details}")
    return "\n".join(lines)

File: internal/render/human.go
package render

import (
	"fmt"
	"strings"

	"github.com/davisbuilds/envdiff/internal/model"
)

func CompareResult(result map[string]any) string {
	lines := []string{
		fmt.Sprintf("Compare %s vs %s", result["left_path"], result["right_path"]),
		fmt.Sprintf("Missing in left: %s", joinStrings(result["missing_in_left"].([]string))),
		fmt.Sprintf("Missing in right: %s", joinStrings(result["missing_in_right"].([]string))),
		fmt.Sprintf("Duplicates in left: %s", joinStrings(result["duplicates_in_left"].([]string))),
		fmt.Sprintf("Duplicates in right: %s", joinStrings(result["duplicates_in_right"].([]string))),
	}

	differing := result["differing_values"].([]map[string]any)
	if len(differing) > 0 {
		lines = append(lines, "Differing values:")
		for _, item := range differing {
			lines = append(
				lines,
				fmt.Sprintf(
					"  %s: %s=%s vs %s=%s",
					item["name"],
					item["left_kind"],
					item["left_value"],
					item["right_kind"],
					item["right_value"],
				),
			)
		}
	}

	return strings.Join(lines, "\n")
}

func MatrixResult(result map[string]any) string {
	lines := []string{
		fmt.Sprintf("Matrix files: %d", result["file_count"]),
		fmt.Sprintf("Variables: %d", result["variable_count"]),
		fmt.Sprintf("Inconsistent: %d", result["inconsistent_variable_count"]),
	}

	warnings := result["warnings"].([]string)
	if len(warnings) > 0 {
		lines = append(lines, fmt.Sprintf("Warnings: %d", len(warnings)))
	}

	variables := result["variables"].([]map[string]any)
	if len(variables) == 0 {
		lines = append(lines, "No variables matched the selected view.")
		return strings.Join(lines, "\n")
	}

	lines = append(lines, "Variables:")
	for _, variable := range variables {
		reasons := []string{}
		if len(variable["missing_in"].([]string)) > 0 {
			reasons = append(reasons, "missing")
		}
		if len(variable["value_kinds"].([]string)) > 1 {
			reasons = append(reasons, "kind-mismatch")
		}
		if len(variable["duplicates_in"].([]string)) > 0 {
			reasons = append(reasons, "duplicate")
		}
		reasonText := strings.Join(reasons, ", ")
		if reasonText == "" {
			reasonText = "consistent"
		}
		lines = append(lines, fmt.Sprintf("  %s [%s]", variable["name"], reasonText))
		for _, fileEntry := range variable["files"].([]map[string]any) {
			if fileEntry["presence"] == "missing" {
				lines = append(lines, fmt.Sprintf("    %s: missing", fileEntry["path"]))
				continue
			}
			suffix := fileEntry["value_kind"].(string)
			if fileEntry["is_duplicate"].(bool) {
				suffix += ", duplicate"
			}
			lines = append(lines, fmt.Sprintf("    %s: present (%s)", fileEntry["path"], suffix))
		}
	}

	return strings.Join(lines, "\n")
}

func ScanResult(scanResult model.RepoScanResult) string {
	lines := []string{
		fmt.Sprintf("Scan root: %s", scanResult.RootPath),
		fmt.Sprintf("Definitions: %d", len(scanResult.Definitions)),
		fmt.Sprintf("Usages: %d", len(scanResult.Usages)),
		fmt.Sprintf("Contracts: %d", len(scanResult.Contracts)),
	}

	if len(scanResult.Contracts) > 0 {
		lines = append(lines, "Contracts:")
		for _, contract := range scanResult.Contracts {
			statuses := strings.Join(contract.Status, ",")
			if statuses == "" {
				statuses = "none"
			}
			lines = append(
				lines,
				fmt.Sprintf("  %s [%s] (%s)", contract.Name, contract.Requiredness, statuses),
			)
		}
	}

	return strings.Join(lines, "\n")
}

func GenerateResult(
	variableCount int,
	outputPath *string,
	annotate bool,
	checkPath *string,
	checkMatches *bool,
) string {
	if checkPath != nil && checkMatches != nil {
		if *checkMatches {
			return fmt.Sprintf("Generated output matches %s", *checkPath)
		}
		return fmt.Sprintf("Generated output drifted from %s", *checkPath)
	}

	suffix := ""
	if annotate {
		suffix = " with annotations"
	}
	if outputPath == nil {
		return fmt.Sprintf("Generated %d variables%s", variableCount, suffix)
	}
	return fmt.Sprintf("Generated %d variables%s to %s", variableCount, suffix, *outputPath)
}

func DoctorResult(
	rootPath string,
	findings []model.Finding,
	suppressedCount int,
	baselineWritten *string,
) string {
	counts := map[string]int{}
	for _, finding := range findings {
		counts[finding.Severity]++
	}
	lines := []string{
		fmt.Sprintf("Doctor root: %s", rootPath),
		fmt.Sprintf("Findings: %d", len(findings)),
		fmt.Sprintf(
			"Summary: %d error, %d warning, %d info",
			counts["error"],
			counts["warning"],
			counts["info"],
		),
	}
	if suppressedCount > 0 {
		lines = append(lines, fmt.Sprintf("Suppressed: %d", suppressedCount))
	}
	if baselineWritten != nil {
		lines = append(lines, fmt.Sprintf("Baseline written: %s", *baselineWritten))
	}
	if len(findings) == 0 {
		lines = append(lines, "No active findings.")
		return strings.Join(lines, "\n")
	}

	for _, severity := range []string{"error", "warning", "info"} {
		scoped := []model.Finding{}
		for _, finding := range findings {
			if finding.Severity == severity {
				scoped = append(scoped, finding)
			}
		}
		if len(scoped) == 0 {
			continue
		}
		lines = append(lines, strings.ToUpper(severity[:1])+severity[1:]+"s:")
		for _, finding := range scoped {
			lines = append(lines, fmt.Sprintf("  %s %s", finding.Code, finding.Details))
		}
	}
	return strings.Join(lines, "\n")
}

func joinStrings(values []string) string {
	if len(values) == 0 {
		return "-"
	}
	return strings.Join(values, ", ")
}
